awsiot/iotshadow.py

```diff
@@ -1323,10 +1338,19 @@ def from_payload(cls, payload):
     def to_payload(self):
         # type: () -> typing.Dict[str, typing.Any]
         payload = {} # type: typing.Dict[str, typing.Any]
-        if self.desired is not None:
+
+        if self.desired_none_is_valid is True:
             payload['desired'] = self.desired
-        if self.reported is not None:
+        else:
+            if self.desired is not None:
+                payload['desired'] = self.desired
```

Comment:
> Set to true to allow reported to be None

your comment says this "allows" the value to be none, but actually it "forces" the value to be none

I would either change the behavior so we that both `desired` needs to be `None` AND `desired_none_is_valid` needs to be `True` before we serialize null.
```
if self.desired is not None:
    payload['desired'] = self.desired
elif self.desired_none_is_valid:
    payload['desired'] = None
```
OR
change the name to `xyz_is_none` and have simpler behavior that just forces use to serialize null, and update the docs
```
if self.desired_is_none:
    payload['desired'] = None
elif self.desired is not None:
   payload['desired'] = self.desired
```

Reply:
It allows it to be `None`, it doesn't force it. If you have `desired_none_is_valid` but `desired` is not `None`, then it will just send like normal. I tested this by setting both `desired_none_is_valid` and `reported_none_is_valid` on the normal shadow message (around line `268`) and it still sent data correctly like expected.
All the code does is skip the `None` check if the `<property>_none_is_valid` property is set to True, so it should still function normally if it has data.

That said, I can adjust it if needed so that, if it's true, it will just send `None`.

Reply:
derp. you're right. I mis-read the code